util/canonical-serializer/src/lib.rs

```diff
@@ -0,0 +1,220 @@
+//! Canonical Serializer
+//! see Readme for details.
+//!
+
+use bytes::Bytes;
+use failure::{ensure, Error};
+use numext_fixed_hash::H256;
+use std::io::Write;
+
+pub type Result<T> = ::std::result::Result<T, Error>;
+
+pub trait CanonicalSerialize {
+    fn serialize<W: Write>(&self, serializer: &mut CanonicalSerializer<W>) -> Result<()>;
+}
+
+pub struct CanonicalSerializer<W> {
+    buf: W,
+}
+
+impl<W: Write> CanonicalSerializer<W> {
+    pub fn new(buf: W) -> Self {
+        CanonicalSerializer { buf }
+    }
+
+    pub fn encode_u8(&mut self, v: u8) -> Result<&mut Self> {
+        self.buf.write_all(&[v])?;
+        Ok(self)
+    }
+
+    pub fn encode_u32(&mut self, v: u32) -> Result<&mut Self> {
+        self.buf.write_all(&v.to_le_bytes())?;
+        Ok(self)
+    }
+
+    pub fn encode_u64(&mut self, v: u64) -> Result<&mut Self> {
+        self.buf.write_all(&v.to_le_bytes())?;
+        Ok(self)
+    }
+
+    // this methods is design for fixed length data
+    pub fn encode_fix_length_bytes(&mut self, v: &[u8], len: usize) -> Result<&mut Self> {
+        ensure!(
+            v.len() == len,
+            "serialize fix length bytes error expect {}, got {}",
+            len,
+            v.len()
+        );
+        self.buf.write_all(v)?;
+        Ok(self)
+    }
+
+    pub fn encode_bytes(&mut self, v: &[u8]) -> Result<&mut Self> {
+        self.encode_u32(v.len() as u32)?;
+        self.buf.write_all(v)?;
+        Ok(self)
+    }
+
+    pub fn encode_vec<T: CanonicalSerialize>(&mut self, list: &[T]) -> Result<&mut Self> {
+        self.encode_u32(list.len() as u32)?;
+        for elem in list {
+            elem.serialize(self)?;
+        }
+        Ok(self)
+    }
+
+    pub fn encode_struct<T: CanonicalSerialize>(&mut self, s: T) -> Result<&mut Self> {
+        s.serialize(self)?;
+        Ok(self)
+    }
+
+    pub fn encode_struct_ref<T: CanonicalSerialize>(&mut self, s: &T) -> Result<&mut Self> {
+        s.serialize(self)?;
+        Ok(self)
+    }
+
+    pub fn encode_option<T: CanonicalSerialize>(&mut self, item: Option<T>) -> Result<&mut Self> {
```

Comment:
I don't think we need this interface. We may need `Option<&T>` in the feature when calculating the hash from flatbuffer or capnp buffer.

Reply:
We need this because of `OutPoint` fields are optional.

Reply:
`encode_option_ref` can cover `Option<OutPoint>`. If we want to keep this, I think we can add a guard `<T: CanonicalSerialize + Copy>`